pantab/_reader.py

```diff
@@ -75,49 +75,69 @@ def _read_table(*, connection: tab_api.Connection, table: TableType) -> pd.DataF
 
 
 def frame_from_hyper(
-    database: Union[str, pathlib.Path],
+    source: Union[str, pathlib.Path, tab_api.Connection],
     *,
     table: TableType,
     hyper_process: Optional[tab_api.HyperProcess] = None,
 ) -> pd.DataFrame:
     """See api.rst for documentation"""
 
-    with tempfile.TemporaryDirectory() as tmp_dir, ensure_hyper_process(
-        hyper_process
-    ) as hpe:
-        tmp_db = shutil.copy(database, tmp_dir)
-        with tab_api.Connection(hpe.endpoint, tmp_db) as connection:
-            return _read_table(connection=connection, table=table)
+    if isinstance(source, tab_api.Connection):
+        forbid_hyper_process(hyper_process)
```

Comment:
Rather than calling this here can you make this a decorator? Would feel a little more Pythonic in nature given how it is used across the functions

Reply:
I don't thik turning this into a decorator would really improve readability.
I would have to duplicate
```
if isinstance(source, tab_api.Connection):
```
both inside the decorator but would still need it in the function body, too.

As such, the code wouldn't become any shorter by turning this into a decorator.
At the same, code flow would be more implicit and thereby harder to follow.

I agree that there are many places where decorators can make code more readable.
But I don't think this is one of those.

Reply:
I see your point on still branching inside of the function body but `forbid_hyper_process` as a decorator would be more natural in Python to do rather than an explicit call within the function

Reply:
I tried to use turn this into a decorator, but I am running into issues around typing.
I am pretty much running into the same issue as https://stackoverflow.com/questions/47060133/python-3-type-hinting-for-decorator and https://github.com/python/mypy/issues/4643. From what I can say, this issue is still unresolved today. 

Afaict, introducing a decorator hence means, we can no longer correctly type `frame_from_hyper` and `frames_from_hyper`. This means tools (linters, autocompletion in IDEs, ...) will have a harder time.

Do we really prefer decorators over having our type hints narrowed down? I would tend towards "type hints are more important", given we should prefer customer benefit (having linters and IDEs work correctly) over pantab-internal considerations (using decorators or not).
 
Or are you maybe aware of a way to work around those typing issues?

Reply:
Hmm OK that is unfortunate. Thanks a lot for the investigation into it. Can clean up at a later date if the typing starts to allow for this